src/librustc/ty/mod.rs

```diff
@@ -256,6 +263,141 @@ impl AssocItem {
     }
 }
 
+/// A list of `ty::AssocItem`s in definition order that allows for efficient lookup by name.
+///
+/// When doing lookup by name, we try to postpone hygienic comparison for as long as possible since
+/// it is relatively expensive. Instead, items are indexed by `Symbol` and hygienic comparison is
+/// done only on items with the same name.
+#[derive(Debug, Clone, PartialEq, HashStable)]
+pub struct AssociatedItems {
+    items_in_def_order: Vec<ty::AssocItem>,
+    item_idx_in_name_order: Vec<usize>,
+}
+
+impl AssociatedItems {
+    /// Constructs an `AssociatedItems` map from a series of `ty::AssocItem`s in definition order.
+    pub fn new(items_in_def_order: Vec<ty::AssocItem>) -> Self {
+        let mut item_idx_in_name_order: Vec<_> =
+            items_in_def_order.iter().enumerate().map(|(i, _)| i).collect();
+
+        // Index the `items_in_def_order` array by name, ignoring hygiene.
+        item_idx_in_name_order.sort_by_key(|&idx| items_in_def_order[idx].ident.name);
+
+        AssociatedItems { items_in_def_order, item_idx_in_name_order }
+    }
+
+    /// Returns a slice of associated items in the order they were defined.
+    ///
+    /// New code should avoid relying on definition order. If you need a particular associated item
+    /// for a known trait, make that trait a lang item instead of indexing this array.
+    //
+    // FIXME(ecstaticmorse): Audit uses of this method to see if the definition order is actually
+    // significant in that context (perhaps we could replace them with calls to a newly defined
+    // `in_arbitrary_order`?). Once all uses have been removed, `AssociatedItems` won't have to
+    // keep a separate sorted list of indexes.
```

Comment:
Can this be accomplished by actually randomizing the order of `items_in_def_order` and see what breaks?